api/types/pam.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	"time"
+
+	"github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/trace"
+)
+
+// PAMConfig represents a PAM configuration resource.
+type PAMConfig interface {
+	// Inherit generic resource requirements.
+	Resource
+
+	// GetEnabled checks if PAM integration should be enabled.
+	GetEnabled() bool
+
+	// SetEnabled configures whether PAM integration should be enabled or not.
+	SetEnabled(bool)
+
+	// GetServiceName returns the PAM service name to use.
+	GetServiceName() string
+
+	// SetServiceName sets the PAM service name to use.
+	SetServiceName(string)
+
+	// GetUsePAMAuth checks if PAM authentication should be used.
+	GetUsePAMAuth() bool
+
+	// SetUsePAMAuth configures whether PAM authentication should be used.
+	SetUsePAMAuth(bool)
+
+	// GetEnvironmont fetches enviroment mappings to set for PAM modules.
+	GetEnvironment() map[string]string
+
+	// SetEnvironment sets environment mappings to set for PAM modules.
+	SetEnvironment(map[string]string)
+
+	// CheckAndSetDefaults configures the resource with default values if empty and validates invariants.
+	CheckAndSetDefaults() error
+}
+
+// NewPAMConfig is a convenience function for creating a PAM configuration from a specification.
+func NewPAMConfig(spec PAMConfigSpecV3) (PAMConfig, error) {
+	conf := PAMConfigV3{
+		Kind:    KindPAMConfig,
+		Version: V3,
+		Metadata: Metadata{
+			Name:      MetaNamePAMConfig,
+			Namespace: defaults.Namespace,
+		},
+		Spec: spec,
+	}
+
+	if err := conf.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &conf, nil
+}
+
+// DefaultPAMConfig creates a default PAM config.
+func DefaultPAMConfig() (PAMConfig, error) {
+	return NewPAMConfig(PAMConfigSpecV3{})
+}
```

Comment:
I don't see this being used, plus it doesn't seem that useful because it would imply a default with everything disabled which is probably not something we want right?

Reply:
Yeah, seems fair, removed.